### DISCUSSION TITLE
讨论：标记选手状态的新字段

### DISCUSSION THREAD

**Discussion by winderica** [UniqueStudio/UniqueRecruitmentBackend#25] [Ideas]
在最初设计的时候，我用`step`字段来表示选手所处的流程这一状态。

但是实际上我忽略了通知短信这个影响因素，考虑以下场景：
1. 场景一（Alice为选手，Bob为某组成员）
    1. Bob 将 Alice 移动到组面流程，但并未发送通知短信；
    2. Bob 过了一会觉得不太行，又把 Alice 移动回了笔试流程；
    3. 只有`step`字段的情况下无法阻止 Alice 在 i 跟 ii 的时间窗口内提交选择组面时间的表单。
    * 目前的解决方式见<https://github.com/UniqueStudio/UniqueRecruitmentBackend/commit/12667fd6a51f510b9767d19ba6cbd89fead16fa0#r48697382>
2. 场景二（Alice为选手，Bob 与 Charlie 为某组成员）
    1. Bob 将 Alice 移动到组面流程并发送短信；
    2. 此时 Charlie 并不知道短信已经发送，他可能：
        1. 再次给 Alice 发送短信， Alice 有点迷惑；
        2. 把 Alice 移动回上一流程，并把她挂了， Alice 超级迷惑。

所以需要有新的字段来标记短信发送的状态，暂时叫`smsStep`吧。

这个字段的行为如下：
1. `step >= smsStep`永远为真，也就是说发送某个流程对应的短信之后，Alice 就无法再被移动到`smsStep`之前的流程；
2. 只有`step > smsStep`时，`step`对应流程的短信才会被成功发送；
3. 校验选手行为/权限的时候使用`smsStep`。

1 与 2 保证了场景二不会发生，而 3 保证了场景一不会发生，且比起目前的解决方案少了一个字段。

上面只是大概的想法，有什么缺陷或者建议欢迎指出。

**Comment by colinaaa** [comment]
我之前也考虑过这个问题，感觉维护两个状态有点复杂（可能会造成误解），我的想法是，在前端可以自由拖动，状态不会同步到其他同学，然后有一个commit的按钮，点击后会**发送短信 & 与同步数据**。

然后对于上面提到的两个场景，可以通过限制*流程只允许正向进行*来避免。

即：
1. 只在commit时发短信
2. 流程只能向前，不能向后

存在的问题：
1. 流程变更和发送短信耦合在一起了，后端可能会变得比较臃肿
2. 拖动选手时没有了同步，多人一起商量时可能体验会变差
3. commit后的数据会同步到其他设备，这时若有未commit的变更就会产生冲突
4. 可能存在操作失误导致的错误通过（例如我们将未通过的选手移到了下一个流程）

优点：
1. 只有一个状态，选手和我们看到的完全一致，简化了使用上的逻辑
2. 简化了现在的手动发送通过短信的流程

**Reply by winderica** [reply]
挂人要怎么办呀，在commit的时候没有拖到下一流程的全挂，这样吗

**Reply by colinaaa** [reply]
挂人也是一种修改，也会做commit，然后在commit的时候发短信呀

**Reply by colinaaa** [reply]
不是吧，那太容易出错了，就跟我们的代码一样，改动了多少就commit多少

**Reply by colinaaa** [reply]
没改动的就还保留原样，只在状态变更的时候通过短信通知选手

**Reply by winderica** [reply]
组面跟群面实际上需要发送两种短信，一种是通知选择时间，另一种是通知面试，也就是面试流程实际上对应了两条短信。如果把流程变更跟发送短信耦合，这两种短信是否发送的状态还是不太好维护吧（不过这么说的话我上面的方案其实也没考虑到

要不增加两个流程，变成：报名->笔试->组面时间选择->组面->熬测->群面时间选择->群面->通过？

**Reply by colinaaa** [reply]
感觉增加流程很不错！可能时间选择那里还可以再细分？

组面时间选择->组面时间待分配->组面？

好像又有些太臃肿了，中间也没有短信需要发送，似乎可以通过 `step === 组面时间选择 && 已选择` 来代表时间待分配

**Reply by winderica** [reply]
ok，那具体再看吧，反正我应该是没书读了，有的是时间（
